tools/ocamlmklib.ml

```diff
@@ -63,15 +63,27 @@ let print_version_num () =
   exit 0;
 ;;
 
+let push_args first arr stk =
+  for i = Array.length arr - 1 downto first do
+    Stack.push arr.(i) stk
+  done
+;;
+
 let parse_arguments argv =
-  let i = ref 1 in
-  let next_arg () =
-    if !i + 1 >= Array.length argv
-    then raise (Bad_argument("Option " ^ argv.(!i) ^ " expects one argument"));
-    incr i; argv.(!i) in
-  while !i < Array.length argv do
-    let s = argv.(!i) in
-    if ends_with s ".cmo" || ends_with s ".cma" then
+  let args = Stack.create () in
+  push_args 1 argv args;
+  let next_arg s =
```

Comment:
You could use `next_arg ()` with `args` as a global stack, and not change the call-sites below.

Reply:
Sorry, I don't understand the suggestion; do you mean to move `next_arg` within the scope of `s` so as not to change the callsites?

Reply:
My comment is bogus because I confused `s` and `args` while writing it. `s` is just used for error messages. You could store it in mutable state on the side, but it's not a good idea, and your current code is fine. Apologies for the confusion.

Reply:
(And declaring it within the scope of `s` would break the spatial locality with the definition of `push_args` that I suggested, unless you move the two below, which is just fairly ugly.)